File: backend/template/tools/social_checker.py
from .base import Tool, ToolType
import asyncio
from typing import Dict, Any
from datetime import datetime, timedelta

class SocialChecker(Tool):
    """ACTIVE tool: Checks social media login status"""
    
    def _determine_type(self) -> ToolType:
        return ToolType.ACTIVE
    
    async def check(self) -> Dict[str, Any]:
        platform = self.config.get("platform", "instagram")
        username = self.config.get("username")
        password = self.config.get("password")
        
        # Check last login (placeholder - would use actual API)
        last_login = await self._check_last_login(platform, username, password)
        monitoring_period_days = self.config.get("monitoring_period_days", 180)
        cutoff_date = datetime.now() - timedelta(days=monitoring_period_days)
        
        trigger = False
        if last_login and last_login < cutoff_date:
            trigger = True
        
        return {
            "status": "ok",
            "data": {
                "platform": platform,
                "username": username,
                "last_login": last_login.isoformat() if last_login else None
            },
            "trigger": trigger
        }
    
    async def run_loop(self, callback):
        self.is_running = True
        interval = self.config.get("interval", 86400)  # Daily check
        while self.is_running:
            result = await self.check()
            if result and result.get("trigger"):
                await callback(result)
            await asyncio.sleep(interval)
    
    async def _check_last_login(self, platform: str, username: str, password: str) -> datetime:
        """Checks last login time (placeholder)"""
        # Placeholder - would use actual social media API
        # For demo, return a date
        return datetime.now() - timedelta(days=30)
    
    def _get_config_schema(self) -> Dict[str, Any]:
        return {
            "platform": {"type": "string", "required": True, "default": "instagram"},
            "username": {"type": "string", "required": True},
            "password": {"type": "string", "required": True},
            "monitoring_period_days": {"type": "int", "default": 180},
            "interval": {"type": "int", "default": 86400}
        }

File: backend/utils/tools_storage/social_checker.ts
/**
 * ACTIVE tool: Checks social media login status
 */
import type { ToolConfig, CheckResult, ConfigSchema } from "./base";

const TOOL_TYPE = "active" as const;

export function getConfigSchema(): ConfigSchema {
  return {
    platform: { type: "string", required: true, default: "instagram" },
    username: { type: "string", required: true },
    password: { type: "string", required: true },
    monitoring_period_days: { type: "int", default: 180 },
    interval: { type: "int", default: 86400 },
  };
}

async function checkLastLogin(
  _platform: string,
  _username: string,
  _password: string
): Promise<Date> {
  // Placeholder - would use actual social media API
  const d = new Date();
  d.setDate(d.getDate() - 30);
  return d;
}

export async function check(config: ToolConfig): Promise<CheckResult> {
  const platform = (config.platform as string) ?? "instagram";
  const username = config.username as string;
  const password = config.password as string;
  const lastLogin = await checkLastLogin(platform, username, password);
  const monitoringPeriodDays = (config.monitoring_period_days as number) ?? 180;
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - monitoringPeriodDays);
  const trigger = lastLogin < cutoffDate;
  return {
    status: "ok",
    data: {
      platform,
      username,
      last_login: lastLogin.toISOString(),
    },
    trigger,
  };
}

export async function runLoop(
  config: ToolConfig,
  callback: (result: CheckResult) => Promise<void> | void
): Promise<void> {
  const interval = (config.interval as number) ?? 86400; // Daily check
  const intervalMs = interval * 1000;
  while (true) {
    const result = await check(config);
    if (result.trigger) {
      await callback(result);
    }
    await new Promise((r) => setTimeout(r, intervalMs));
  }
}
